File: Tic Toe/Game.ts
import { Board } from "./Models/Board";
import { Player } from "./Models/Player";
import { PlayingPieceO } from "./Models/PlayingPieceO";
import { PlayingPieceX } from "./Models/PlayingPieceX";

export class Game {
  public board: Board;
  public player1: Player;
  public player2: Player;
  constructor() {
    this.board = new Board();
    const pieceX = new PlayingPieceX();
    const pieceO = new PlayingPieceO();
    this.player1 = new Player("Player 1", pieceX);
    this.player2 = new Player("Player 2", pieceO);
  }

  play() {
    const currentPlayer: Player[] = [];
    currentPlayer.push(this.player1, this.player2);

    while (1) {
      this.board.printBoard();

      if (!this.board.isFree()) {
        break;
      }

      const player = currentPlayer[0];

      currentPlayer.shift();
      currentPlayer.push(player);
      console.log(player.getName() + "'s turn");

      const row = parseInt(prompt("Enter row: ") as string);
      const col = parseInt(prompt("Enter col: ") as string);

      if (this.board.checkFree(row, col)) {
        this.board.placePiece(row, col, player.getPiece().piece);
        if (this.board.checkWinner(player.getPiece().piece)) {
          this.board.printBoard();
          console.log(player.getName() + " wins!");
          break;
        }
      } else {
        console.log("Invalid move");
        continue;
      }
    }
  }
}

File: TicToe/Game.java
package TicToe;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Game {

    public static void main(String[] args) {

        PlayingPiece pieceX = new PlayingPieceX();
        PlayingPiece pieceO = new PlayingPieceO();

        Player player1 = new Player("Player 1", pieceX);
        Player player2 = new Player("Player 2", pieceO);

        Board board = new Board();

        Queue<Player> queue = new LinkedList<Player>();
        queue.add(player1);
        queue.add(player2);

        Scanner sc = new Scanner(System.in);

        while (true) {
            Player player = queue.poll();
            board.printBoard();

            System.out.println("Enter q to quit or press enter to continue");
            String input = sc.nextLine();
            if (input.equals("q")) {
                System.out.println("Game Over");
                System.out.println(player.getName() + "Score: " + player.getScore());
                System.out.println(player2.getName() + "Score: " + player2.getScore());
                break;
            }

            System.out.println("Player " + player.getName() + " turn");
            System.out.println("Enter row: ");
            int row = sc.nextInt();

            System.out.println("Enter col: ");
            int col = sc.nextInt();

            if (board.getPosition(row, col) == ".") {
                board.setPosition(row, col, player.getPiece().getPiece().toString());

                if (board.checkWinner(player.getPiece().getPiece().toString())) {
                    board.printBoard();
                    System.out.println(player.getName() + " wins!");
                    player.incrementScore();
                    board.clearBoard();
                }

                queue.add(player);
            } else {
                System.out.println("Invalid move");
                continue;
            }
        }

        sc.close();
    }
}
